pkg/fleets/controller.go

```diff
@@ -514,19 +514,21 @@ func (c *Controller) cleanupUnhealthyReplicas(ctx context.Context, rest []*agone
 	// Safely scale down all old GameServerSets with unhealthy replicas.
 	totalScaledDown := int32(0)
 	for i, gsSet := range rest {
+		healthy := gsSet.Status.ReadyReplicas + gsSet.Status.AllocatedReplicas + gsSet.Status.ReservedReplicas
```

Comment:
Thanks for digging into this so deeply!

I probably need to sit down and play with this to see it in effect, but I would definitely want this to sit behind a feature gate, so it can be widely tested without fear of breaking things.

To do that, my suggestion would be is duplicate the entirety of `cleanupUnhealthyReplicas` and `rollingUpdateRestFixedOnReady` with new names, e.g. "cleanupUnhealthyReplicasRollingUpdateFix` (or something similar).

(Maybe in a new file, if the controller is getting messy)

Then we could do something akin to the following:

```go
func (c *Controller) cleanupUnhealthyReplicas(ctx context.Context, rest []*agonesv1.GameServerSet,
	fleet *agonesv1.Fleet, maxCleanupCount int32) ([]*agonesv1.GameServerSet, int32, error) {

  if runtime.FeatureEnabled(runtime.FeatureRollingUpdateFix) {
    return cleanupUnhealthyReplicasRollingUpdateFix(ctx, rest, fleet, maxCleanupCount)
  }

// ....
}
```

And we could turn on/off feature flags and so we could have relavent unit tests as well.

How does that sound?

Reply:
Sure, as you wish.

Reply:
I have managed to prove the primary tenants of the fix in unit tests, which is much simpler than e2e.

As a side note, `TestControllerRollingUpdateDeployment` seems to break the laws under with GSS Status Replicas exists, but I left the original math unchanged.